staging/src/k8s.io/apiserver/pkg/util/proxy/websocket.go

```diff
@@ -0,0 +1,192 @@
+func createChannels(opts Options) []wsstream.ChannelType {
+	// open the requested channels, and always open the error channel
+	channels := make([]wsstream.ChannelType, 5)
+	channels[constants.StreamStdIn] = readChannel(opts.Stdin)
+	channels[constants.StreamStdOut] = writeChannel(opts.Stdout)
+	channels[constants.StreamStdErr] = writeChannel(opts.Stderr)
+	channels[constants.StreamErr] = wsstream.WriteChannel
+	channels[constants.StreamResize] = wsstream.ReadChannel
+	return channels
+}
+
+// readChannel returns wsstream.ReadChannel if real is true, or wsstream.IgnoreChannel.
+func readChannel(real bool) wsstream.ChannelType {
+	if real {
+		return wsstream.ReadChannel
+	}
+	return wsstream.IgnoreChannel
+}
+
+// writeChannel returns wsstream.WriteChannel if real is true, or wsstream.IgnoreChannel.
+func writeChannel(real bool) wsstream.ChannelType {
+	if real {
+		return wsstream.WriteChannel
+	}
+	return wsstream.IgnoreChannel
+}
+
+// createWebSocketStreams returns a "conns" struct containing the websocket connection and
+// streams needed to perform an exec or an attach.
+func createWebSocketStreams(req *http.Request, w http.ResponseWriter, opts Options) (*conns, error) {
+	channels := createChannels(opts)
+	conn := wsstream.NewConn(map[string]wsstream.ChannelProtocolConfig{
+		// WebSocket server only supports remote command version 5.
+		constants.StreamProtocolV5Name: {
+			Binary:   true,
+			Channels: channels,
+		},
+	})
+	conn.SetIdleTimeout(defaultIdleConnectionTimeout)
+	// Opening the connection responds to WebSocket client, negotiating
+	// the WebSocket upgrade connection and the subprotocol.
+	negotiatedProtocol, streams, err := conn.Open(w, req)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("unable to upgrade websocket connection: %v", err))
+		return nil, err
+	}
+	klog.V(4).Infof("websocket server negotiated sub-protocol: %s", negotiatedProtocol)
+
+	// Send an empty message to the lowest writable channel to notify the client the connection is established
+	switch {
+	case opts.Stdout:
+		_, err = streams[constants.StreamStdOut].Write([]byte{})
+	case opts.Stderr:
+		_, err = streams[constants.StreamStdErr].Write([]byte{})
+	default:
+		_, err = streams[constants.StreamErr].Write([]byte{})
+	}
+	if err != nil {
+		return nil, fmt.Errorf("write error during websocket server creation: %v", err)
+	}
+
+	ctx := &conns{
+		conn:         conn,
+		stdinStream:  streams[constants.StreamStdIn],
+		stdoutStream: streams[constants.StreamStdOut],
+		stderrStream: streams[constants.StreamStdErr],
+		tty:          opts.Tty,
+		resizeStream: streams[constants.StreamResize],
+	}
+
+	// writeStatus returns a WriteStatusFunc that marshals a given api Status
+	// as json in the error channel.
+	ctx.writeStatus = func(status *apierrors.StatusError) error {
+		bs, err := json.Marshal(status.Status())
+		if err != nil {
+			return err
+		}
+		// Write status error to error stream with deadline.
+		conn.SetWriteDeadline(writeErrorDeadline)
```

Comment:
is SetWriteDeadline threadsafe, and if not, what ensures writeStatus() will not be called in parallel with other calls?

in stream#Write, we use `s.connWriteLock` but we don't have that here, right?

Reply:
We're using the gorilla websockets library in the client (which is where `s.connWriteLock` lives). But the websocket server is based on existing code (in `kubelet` and the container runtimes) which uses the golang standard websocket library. This websocket library says
```
Multiple goroutines may invoke methods on a Conn simultaneously.
```
https://pkg.go.dev/golang.org/x/net/websocket#Conn